package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py

```diff
@@ -679,90 +679,136 @@ def _get_dh_pairs(self):
 
         return donors, hydrogens
 
-    def _filter_atoms(self, donors, acceptors):
-        """Create a mask to filter donor, hydrogen and acceptor atoms.
-
-        This can be used to consider only hydrogen bonds between two or more
-        specified groups.
-
-        Groups are specified with the `between` keyword when creating the
-        HydrogenBondAnalysis object.
-
-           Returns
-           -------
-           mask: np.ndarray
-
-
-        .. versionchanged:: 2.5.0
-           Change return value to a mask instead of separate AtomGroups.
-``
-        """
-
-        mask = np.full(donors.n_atoms, fill_value=False)
-        for group1, group2 in self.between_ags:
-
-            # Find donors in G1 and acceptors in G2
-            mask[
-                    np.logical_and(
-                        np.isin(donors.indices, group1.indices),
-                        np.isin(acceptors.indices, group2.indices)
-                    )
-            ] = True
-
-            # Find acceptors in G1 and donors in G2
-            mask[
-                np.logical_and(
-                    np.isin(acceptors.indices, group1.indices),
-                    np.isin(donors.indices, group2.indices)
-                )
-            ] = True
-
-        return mask
-
-
     def _prepare(self):
         self.results.hbonds = [[], [], [], [], [], []]
-
-
+ 
     def _single_frame(self):
-
         box = self._ts.dimensions
 
         # Update donor-hydrogen pairs if necessary
         if self.update_selections:
             self._donors, self._hydrogens = self._get_dh_pairs()
 
-        # find D and A within cutoff distance of one another
-        # min_cutoff = 1.0 as an atom cannot form a hydrogen bond with itself
-        d_a_indices, d_a_distances = capped_distance(
-            self._donors.positions,
-            self._acceptors.positions,
-            max_cutoff=self.d_a_cutoff,
-            min_cutoff=1.0,
-            box=box,
-            return_distances=True,
-        )
+        if self.between_ags is not None:
+            # map donor indices to respective hydrogen indices
+            donor_idx_to_H_indices = defaultdict(list)
+
+            for d, h in zip(self._donors, self._hydrogens):
+                donor_idx_to_H_indices[d.index].append(h.index)
+
+            tmp_hbond_candidates = [] 
+
+            for group1, group2 in self.between_ags:
+
+                tmp_donors_g1_mask = np.isin(self._donors.indices, group1.indices) 
+                donors_for_dist = self._donors[tmp_donors_g1_mask] 
+                tmp_acceptors_g2_mask = np.isin(self._acceptors.indices, group2.indices) 
+                acceptors_for_dist = self._acceptors[tmp_acceptors_g2_mask] 
```

Comment:
```python
                tmp_donors_mask = np.logical_or(
                    np.isin(self._donors.indices, group1.indices),
                    np.isin(self._donors.indices, group2.indices),
                 )
                donors_for_dist = self._donors[tmp_donors_mask] 
                tmp_acceptors_mask = np.logical_or(
                    np.isin(self._acceptors.indices, group1.indices),
                    np.isin(self._acceptors.indices, group2.indices),
                ) 
                acceptors_for_dist = self._acceptors[tmp_acceptors_mask] 
```
I think if you do this you only need to make one call to `capped_distance` per pair of groups, rather than calculating distances between g1 donors to g2 acceptors and then g2 donors to g1 donors